### DISCUSSION TITLE
Optimal workflow for extension development?

### DISCUSSION THREAD

**Discussion by CharlieDigital** [directus/directus#14319] [Archive]
Trying to figure out an optimal flow for fast DX based on [the documentation on extension development](https://docs.directus.io/extensions/creating-extensions/).

I am using `docker compose` and have mounted a local `.directus/extensions` directory to the container:

```
  volumes:
      - ./.directus/extensions:/directus/extensions
```
My directory structure is like so:

```
/.directus
  /data
  /extensions
    /displays
      /my-custom-display
        -index.js
  /uploads
/src
  /my-custom-display
    /dist
      -index.js
    /src
      -display.vue 
      -index.ts
```

I noticed that there is a `--watch` flag.

Of course, this will watch for file changes in my extension and rebuild, but by default pushes the output to `./dist` which then has to be copied to the `./.directus/extensions/displays/my-custom-display/` directory.

```
cd src/my-custom-display
npm run build -- --watch

# Changes will produce ./dist/index.js; copy to the mounted extension directory

cp ./dist/*.* ../../.directus/extensions/displays/my-custom-display
```

I'm wondering what the optimal flow for extension development is.

## Symbolic Link

[I saw one repo using a symbolic link approach](https://github.com/wellenplan/wellenplan/wiki/Development):

```
cd ./.directus/extensions/displays/my-custom-display/
ln -s ../../../.../src/my-custom-display/dist/index.js index.js
```
The symlink is created, but Directus shows an error on load: 

```
directus  | 14:40:15 ✨ Loaded extensions: my-custom-display
directus  | 14:40:15 ✨ Initializing bootstrap...
directus  | 14:40:16 ✨ Database already initialized, skipping install
directus  | 14:40:16 ✨ Running migrations...
directus  | 14:40:16 ✨ Done
directus  | 14:40:17 ⚠️  Couldn't bundle App extensions
directus  | 14:40:17 ⚠️  Could not resolve './extensions/displays/my-custom-display/index.js' from virtual:entry
directus  | 14:40:17 ✨ Loaded extensions: my-custom-display
```

This approach seems like it would be great, but it doesn't seem to be supported or I'm missing some other detail/configuration.

## Modify `package.json`

By default, the `package.json` has the following:

```json
	"directus:extension": {
		"type": "display",
		"path": "dist/index.js",       # Change this to output to the mounted volume
		"source": "src/index.ts",
		"host": "^9.13.0"
	},
```

I can change this:

```json
	"directus:extension": {
		"type": "display",
		"path": "../../.directus/extensions/displays/my-custom-display/index.js",
		"source": "src/index.ts",
		"host": "^9.13.0"
	},
```

Then in `docker-compose.yaml`:

```
  environment:
    EXTENSIONS_AUTO_RELOAD: true
```

This works, but the problem is that in a CI build, I'd want to package from `/dist` so it seems that some mechanism that preserves `/dist` as the target would be best.

## Questions

1. Is there an optimal flow that makes cycling and DX faster?
2. Have I missed something with the symlink approach?
3. While `EXTENSIONS_AUTO_RELOAD` does indeed reload the extension in the runtime, I need to manually refresh the page in the browser; is there a possibility for hot-reload of Directus itself with the `docker-compose` container?  Or would I need to build and run from source for that?

**Comment by br41nslug** [comment]
1) I don't think one approach is really faster than the other but i usually go the "Modify package.json" route.
2) Symlinks are tricky when running in docker, i suspect you have made the symlink on the host not inside the container which won't work.
Maybe a compromise is to keep building to `/dist` but solve the location problem in how you mount the extensions into the docker, you could do something like 
``` 
volumes:
      - ./src/my-custom-display/dist:/directus/extensions/displays/my-custom-display
```
3) no hot reloading of frontend extension is currently not possible

**Reply by CharlieDigital** [reply]
Thanks for the feedback; point 2 is a great idea.
